lineax/_solver/lsmr.py

```diff
@@ -0,0 +1,583 @@
+        def early_stop_normb(loop_state):
+            return (
+                jtu.tree_map(jnp.zeros_like, operator.in_structure()),
+                istop,
+                itn,
+                normr,
+                normar,
+                normA,
+                condA,
+                self.norm(x),
+            )
+
+        def no_early_stop(loop_state):
+            loop_state = lax.while_loop(condfun, bodyfun, loop_state)
+
+            (
+                loop_state_main,
+                loop_state_r_est,
+                loop_state_anorm,
+                loop_state_stopping,
+                loop_state_vecs,
+            ) = loop_state
+            (
+                itn,
+                alpha,
+                beta,
+                zetabar,
+                alphabar,
+                rho,
+                rhobar,
+                cbar,
+                sbar,
+            ) = loop_state_main
+            (
+                betadd,
+                betad,
+                rhodold,
+                tautildeold,
+                thetatilde,
+                zeta,
+                d,
+            ) = loop_state_r_est
+            (normA2, maxrbar, minrbar, normA, condA) = loop_state_anorm
+            (istop, normr, normb) = loop_state_stopping
+            (x, u, v, h, hbar) = loop_state_vecs
+
+            normar = abs(zetabar)
+            return x, istop, itn, normr, normar, normA, condA, self.norm(x)
+
+        def early_stop(loop_state):
+            return lax.cond(
+                normar == 0, early_stop_normar, early_stop_normb, loop_state
+            )
+
+        x, istop, itn, normr, normar, normA, condA, normx = lax.cond(
+            jnp.logical_or(normar == 0, normb == 0),
+            early_stop,
+            no_early_stop,
+            loop_state,
+        )
+
+        stats = {
+            "num_steps": itn,
+            "norm_r": normr,
+            "norm_Ar": normar,
+            "norm_A": normA,
+            "cond_A": condA,
+            "norm_x": normx,
+        }
+
+        # TODO: return actual status when failed
+        result = RESULTS.where(
+            istop == 0,
+            RESULTS.successful,
+            RESULTS.max_steps_reached,
+        )
+
+        return x, result, stats
+
+    def _givens(self, a, b):
+        """Stable implementation of Givens rotation, from [1]_
+
+        finds c, s, r such that
+
+        |c  -s|[a| = |r|
+        [s   c|[b|   |0|
+
+        r = sqrt(a^2 + b^2)
+
+        Assumes a, b are real.
+
+        References
+        ----------
+        .. [1] S.-C. Choi, "Iterative Methods for Singular Linear Equations
+            and Least-Squares Problems", Dissertation,
+            http://www.stanford.edu/group/SOL/dissertations/sou-cheng-choi-thesis.pdf
+
+        """
+
+        def bzero(a, b):
+            return jnp.sign(a), 0.0, abs(a)
+
+        def azero(a, b):
+            return 0.0, jnp.sign(b), abs(b)
+
+        def b_gt_a(a, b):
+            tau = a / b
```

Comment:
I suspect this could do something funny for the same reason as above -- we may actually have `a = 0` here.
I think whenever dividing by `a` or `b` it should probably be replaced with something like `/ jnp.where(b==0, 1, b)` (and then the outer `cond` will ensure that the `b=0` case is still handled correctly, by just ignoring the results of this computation).